ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -4230,6 +4229,34 @@ public static long unsetBit(long bitmap, int bitIdx) {
     }
   }
 
+  protected boolean isGroupBy(ASTNode expr) {
+    boolean isGroupBy = false;
+    if (expr.getParent() != null && expr.getParent() instanceof Node)
+    for (Node sibling : ((Node)expr.getParent()).getChildren()) {
+      isGroupBy |= sibling instanceof ASTNode && ((ASTNode)sibling).getType() == HiveParser.TOK_GROUPBY;
+    }
+
+    return isGroupBy;
+  }
+
+  protected boolean isSelectDistinct(ASTNode expr) {
+    return expr.getType() == HiveParser.TOK_SELECTDI;
+  }
+
+  protected boolean isAggregateInSelect(Node node, Collection<ASTNode> aggregateFunction) {
+    if (node.getChildren() == null) {
+      return false;
+    }
+
+    for (Node child : node.getChildren()) {
```

Comment:
is it safe to traverse all the children? I'm wondering if can't we can get a false + somehow...

Reply:
I don't see how, all the nodes under the SELECT DISTINCT node are for expressions that may be aggregations. Under the SELECT DISTINCT node there may be things like "1 + count(*)" in it, in which case the aggregation is at a lower level in the tree.

Reply:
I was thinking something really odd:
```
select distinct (select count(*) from t where t.a=e.a) from e
```
but in this case (beyond that it might not accept by hive at all) the count aggregate is not present at the top level.
Do you know an example when this method returns false; however there are aggreagations being done?

Reply:
I doubt there is any. The above example is not valid, it says:

Unsupported SubQuery Expression Invalid subquery. Subquery with DISTINCT clause is not supported!